### ISSUE TITLE
[Perf regression] Converting storage to use tonic reduced TPS by 15%

### ISSUE THREAD

**Issue by andll** [libra/libra#2213]
Converting storage client to tonic seem to reduce cluster test reported performance from 650 to 550 TPS

**Comment by wqfish** [comment]
The [mutex](https://github.com/libra/libra/blob/d9e5e95178091df539b8c93884fc25a3b6756766/storage/storage-client/src/lib.rs#L45) looks a bit suspicious to me. Does that cause some kind of contention and @bmwill do we really need a mutex?

**Comment by bmwill** [comment]
Couple things:
* The Mutex is only used because tonic's API is different from grpcio's in that all client methods take an `&mut` ref in order to make RPC calls.  Since we are using `Arc<dyn T>` everywhere (which we shouldn't be IMO) we can't get an `&mut` ref to the client to actually make the call, hence the Mutex and cloning the client on each call (the clone is actually really cheap though). There may be contention on the mutex but its only held for such a brief period of time that it may not be a problem. Either way the best way forward on this front is to stop using `Arc<dyn T>` everywhere and instead of sharing clients with multiple tasks, just have a client that each task owns.
* We used to have a Vec<Client> and this seemed like we only had this because of some limitations in grpcio's impl (and because we were sharing clients with multiple threads) so I removed that. The solution here is also to probably stop sharing clients and instead have an owned client per place we need one (i.e. get rid of `Arc<dyn T>`)

Last endpoint I converted I went through the hassle of converting all the `Arc<dyn T>` mess with mempool but I got some backlash for changing a trait in consensus. So this time around I figured I'd find a way to do the conversion without dramatically changing the APIs first and then at a later point we could go and make the appropriate change.

**Comment by wqfish** [comment]
In order to understand the perf impact of tonic I did some experiments. I made a change to kill the gRPC layer entirely -- every storage read client becomes `Arc<LibraDB>`. There is no RPC call to storage service any more and everything is direct function call. This could give an upper bound of how much we can get from optimizing the gRPC libraries.

- Without the change the numbers on master are like `Experiment Result: all up : 635 TPS, 1026.5 ms latency`.
- With the change the numbers are like `Experiment Result: all up : 746 TPS, 448.8 ms latency`. I also observed some discrepancy between the final number reported in the terminal and the grafana dashboard. From the dashboard the TPS looks more like 900~1000. Not sure which one is more accurate.

![Screenshot 2020-01-24 13 16 44](https://user-images.githubusercontent.com/1171005/73104681-d4476580-3eab-11ea-9571-33ea362b323c.png)

So it looks like the gRPC layer might have some significant impact on the overall perf, especially on the latency numbers.

If anyone wants to play with the change, the code is here: https://github.com/libra/libra/pull/2283
